code/modules/holodeck/computer.dm

```diff
@@ -110,207 +146,268 @@
+	//holo effects are taken out of the spawned list and added to the effects list
+	//turfs and overlay objects are taken out of the spawned list
+	//objects get resistance flags added to them
+	for (var/_atom in spawned)
+		var/atom/atoms = _atom
+
+		if (isturf(atoms) || istype(atoms, /obj/effect/overlay/vis)) //ssatoms
+			spawned -= atoms
+			continue
+
+		atoms.flags_1 |= HOLOGRAM_1
+		RegisterSignal(atoms, COMSIG_PARENT_PREQDELETED, .proc/remove_from_holo_lists)
+
+		if (isholoeffect(atoms))//activates holo effects and transfers them from the spawned list into the effects list
+			var/obj/effect/holodeck_effect/holo_effect = atoms
+			effects += holo_effect
+			spawned -= holo_effect
+			var/atom/active_effect = holo_effect.activate(src)
+			if(istype(active_effect) || islist(active_effect))
+				spawned += active_effect // we want mobs or objects spawned via holoeffects to be tracked as objects
+			continue
+
+		if (isobj(atoms))
+			var/obj/holo_object = atoms
+			holo_object.resistance_flags = LAVA_PROOF | FIRE_PROOF | UNACIDABLE | ACID_PROOF
+
+			if (isstructure(holo_object))
+				holo_object.flags_1 |= NODECONSTRUCT_1
+				continue
+
+			if (ismachinery(holo_object))
+				var/obj/machinery/machines = holo_object
+				machines.flags_1 |= NODECONSTRUCT_1
+				machines.power_change()
+
+				if(istype(machines, /obj/machinery/button))
+					var/obj/machinery/button/buttons = machines
+					buttons.setup_device()
+
+///this qdels holoitems that should no longer exist for whatever reason
+/obj/machinery/computer/holodeck/proc/derez(obj/object, silent = TRUE, forced = FALSE)
+	if(!object)
+		return
+
+	spawned -= object
+	UnregisterSignal(object, COMSIG_PARENT_PREQDELETED)
+	var/turf/target_turf = get_turf(object)
+	for(var/c in object) //make sure that things inside of a holoitem are moved outside before destroying it
+		var/atom/movable/object_contents = c
+		object_contents.forceMove(target_turf)
+
+	if(!silent)
+		visible_message("<span class='notice'>[object] fades away!</span>")
+
+	qdel(object)
+
+/obj/machinery/computer/holodeck/proc/remove_from_holo_lists(datum/to_remove, _forced)
+	spawned -= to_remove
+	UnregisterSignal(to_remove, COMSIG_PARENT_PREQDELETED)
 
 /obj/machinery/computer/holodeck/process(delta_time)
 	if(damaged && DT_PROB(5, delta_time))
-		for(var/turf/T in linked)
+		for(var/turf/holo_turf in linked)
 			if(DT_PROB(2.5, delta_time))
-				do_sparks(2, 1, T)
+				do_sparks(2, 1, holo_turf)
 				return
-
-	if(!..() || !active)
+	. = ..()
+	if(!. || program == offline_program)//we dont need to scan the holodeck if the holodeck is offline
 		return
 
-	if(!floorcheck())
+	if(!floorcheck()) //if any turfs in the floor of the holodeck are broken
 		emergency_shutdown()
 		damaged = TRUE
-		for(var/mob/M in urange(10,src))
-			M.show_message("The holodeck overloads!")
+		for(var/mob/can_see_fuckup in urange(10,src))
+			can_see_fuckup.show_message("The holodeck overloads!")
```

Comment:
Is this a `visible_message()`?

Reply:
old holodeck code, its now a visible message